### ISSUE TITLE
traverse: port `scope.push` API from Babel

### ISSUE THREAD

**Issue by Dunqing** [oxc-project/oxc#5049]
Babel's implementation is [here](https://github.com/babel/babel/blob/14308d35ce77b4bac30ecd37aec817046402c455/packages/babel-traverse/src/scope/index.ts#L1075-L1144)

### Why need this API?

We need to combine the variable declarators required by different plugins to be inserted at the top of the statements.

### Related tests:

* logical-assignment/null-coalescing/input.js

```diff
-var _obj$x, _obj$x2, _deep$obj, _deep$obj$x, _deep$obj2, _deep$obj2$x, _key, _obj$_key, _key2, _obj$_key2, _deep$obj3, _key3, _deep$obj3$_key, _deep$obj4, _key4, _deep$obj4$_key;
+var _obj$x, _obj$x2, _deep$obj$x, _deep$obj2$x, _obj$_key, _obj$_key2, _deep$obj3$_key, _deep$obj4$_key;
+var _deep$obj, _deep$obj2, _key, _key2, _deep$obj3, _key3, _deep$obj4, _key4;
```

**Comment by Dunqing** [comment]
When we have this API, we can remove the following code. This implementation appears in many plugins
https://github.com/oxc-project/oxc/blob/200d887852e50a252276b1c2009890a987e2be5f/crates/oxc_transformer/src/es2021/logical_assignment_operators.rs#L79-L100

**Comment by overlookmotel** [comment]
Yes, agreed we need this. Aside from the failing test, the repeated `exit_statements` pattern is inefficient.

I'm not sure which is best way to implement it though. Ideally it wouldn't be a method on `TraverseCtx` because it's not really related to traversal. But it's shared state between different transforms, which we're don't have a facility for at present.

By the way, Babel's implementation is inefficient:

```js
// Input
{ obj.a &&= 1; }
{ obj.b &&= 2; }
```

```js
// Output
{
  var _obj;
  (_obj = obj).a && (_obj.a = 1);
}
{
  var _obj2;
  (_obj2 = obj).b && (_obj2.b = 2);
}
```

[Babel REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=N4KABBYPYEYFYDoCGYBkqC8YCMBuEAviKJNPAjGpmAEz5FA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=react%2Cstage-2&prettier=false&targets=&version=7.25.3&externalPlugins=babel-plugin-transform-es2015-destructuring%406.23.0&assumptions=%7B%7D)

But as they are `var` statements, they could be hoisted up and combined:

```js
// Output
var _obj, _obj2;
{ (_obj = obj).a && (_obj.a = 1); }
{ (_obj2 = obj).b && (_obj2.b = 2); }
```

This is more compact output (though minifier would probably do that hoisting anyway) but also requires maintaining an entry in `var_declarations` for every block, when it could be just be top level / function blocks.

**Comment by overlookmotel** [comment]
We can use `SparseStack` (introduced in #5940).

The problem with putting this in `TransformCtx` is that then it needs to be wrapped in a `RefCell` which will be somewhat costly, as it's accessed often.

Maybe should parameterize `TraverseCtx` so it can hold arbitrary state:

```rs
pub trait Traverse<'a, State> {
    fn enter_statements(&mut self, stmts: &mut Vec<'a, Statement<'a>>, ctx: &mut TraverseCtx<'a, State>) {}
    fn exit_statements(&mut self, stmts: &mut Vec<'a, Statement<'a>>, ctx: &mut TraverseCtx<'a, State>) {}
    // etc...
}

pub struct TraverseCtx<'a, State> {
    pub ancestry: TraverseAncestry<'a>,
    pub scoping: TraverseScoping,
    pub ast: AstBuilder<'a>,
    pub state: State,
}

struct TransformerState<'a> {
    var_declarations: SparseStack<Vec<'a, VariableDeclarator<'a>>>,
}

type TransCtx<'a> = TraverseCtx<'a, TransformerState<'a>>;

struct CommonTransform;

impl<'a> Traverse<'a, TransformerState<'a>> for CommonTransform {
    fn enter_statements(&mut self, _stmts: &mut Vec<'a, Statement<'a>>, ctx: &mut TransCtx<'a>) {
        ctx.state.var_declarations.push(None);
    }

    fn exit_statements(&mut self, stmts: &mut Vec<'a, Statement<'a>>, ctx: &mut TransCtx<'a>) {
        if let Some(declarations) = ctx.state.var_declarations.pop() {
            // Insert `var` declaration statement
        }
    }
}
```

To do:

1. Put it in `TransformCtx` first, in a `RefCell`.
2. Move to `TraverseCtx` and see if it makes a difference to perf.

**Comment by overlookmotel** [comment]
#6170 fixed the failing test mentioned at top of this issue.

Only thing remaining is for arrow functions transform to also use `VarDeclarations` for inserting `var _this = this;` statements.

**Comment by overlookmotel** [comment]
@Dunqing I think arrow functions transform now uses `VarDeclarations` for adding statements, right? If so, I think we can close this issue.

**Comment by Dunqing** [comment]
Yes, It is done!
